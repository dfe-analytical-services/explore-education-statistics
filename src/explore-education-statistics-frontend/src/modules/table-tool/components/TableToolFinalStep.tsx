import Tag from '@common/components/Tag';
import useAsyncRetry from '@common/hooks/useAsyncRetry';
import useToggle from '@common/hooks/useToggle';
import DownloadTable from '@common/modules/table-tool/components/DownloadTable';
import TableHeadersForm from '@common/modules/table-tool/components/TableHeadersForm';
import TimePeriodDataTable from '@common/modules/table-tool/components/TimePeriodDataTable';
import { SelectedPublication } from '@common/modules/table-tool/types/selectedPublication';
import TableToolShare from '@frontend/modules/table-tool/components/TableToolShare';
import TableToolInfo from '@common/modules/table-tool/components/TableToolInfo';
import { FullTable } from '@common/modules/table-tool/types/fullTable';
import { TableHeadersConfig } from '@common/modules/table-tool/types/tableHeaders';
import publicationService from '@common/services/publicationService';
import Link from '@frontend/components/Link';
import tableBuilderService, {
  TableDataQuery,
} from '@common/services/tableBuilderService';
import { logEvent } from '@frontend/services/googleAnalyticsService';
import React, { memo, ReactNode, useRef } from 'react';

interface TableToolFinalStepProps {
  query: TableDataQuery;
  table: FullTable;
  tableHeaders: TableHeadersConfig;
  selectedPublication: SelectedPublication;
  newPermalinks?: boolean; // TO DO - EES-4259 remove `newPermalinks` param and tidy up
  onReorderTableHeaders: (reorderedTableHeaders: TableHeadersConfig) => void;
}

const TableToolFinalStep = ({
  table,
  tableHeaders,
  query,
  selectedPublication,
  newPermalinks,
  onReorderTableHeaders,
}: TableToolFinalStepProps) => {
  const dataTableRef = useRef<HTMLElement>(null);
  const [hasTableError, toggleHasTableError] = useToggle(false);

  const { value: fullPublication } = useAsyncRetry(
    async () =>
      publicationService.getLatestPublicationRelease(selectedPublication.slug),
    [selectedPublication],
  );
  const publication = fullPublication?.publication;

  const getMethodologyLinks = () => {
    const links: ReactNode[] =
      publication?.methodologies?.map(methodology => (
        <Link key={methodology.id} to={`/methodology/${methodology.slug}`}>
          {methodology.title}
        </Link>
      )) ?? [];

    if (publication?.externalMethodology) {
      links.push(
        <Link
          key={publication.externalMethodology.url}
          to={publication.externalMethodology.url}
        >
          {publication.externalMethodology.title}
        </Link>,
      );
    }
    return links;
  };

  return (
    <div
      className="govuk-!-margin-bottom-4"
      data-testid="Table tool final step container"
    >
      {table && tableHeaders && (
        <>
          <div className="govuk-!-margin-bottom-3 dfe-flex dfe-align-items-start dfe-justify-content--space-between">
            {selectedPublication.selectedRelease.latestData && (
              <Tag strong>This is the latest data</Tag>
            )}

            {!selectedPublication.selectedRelease.latestData && (
              <>
                <div className="govuk-!-margin-bottom-3">
                  <Tag strong colour="orange">
                    This data is not from the latest release
                  </Tag>
                </div>

                <Link
                  className="dfe-print-hidden"
                  unvisited
                  to={`/find-statistics/${selectedPublication.slug}`}
                  testId="View latest data link"
                >
                  View latest data:{' '}
                  <span className="govuk-!-font-weight-bold">
                    {selectedPublication.latestRelease.title}
                  </span>
                </Link>
              </>
            )}
          </div>

          <TableHeadersForm
            initialValues={tableHeaders}
            onSubmit={nextTableHeaders => {
              onReorderTableHeaders(nextTableHeaders);
              if (dataTableRef.current) {
                // add a short delay so the reordering form is closed before it scrolls.
                setTimeout(() => {
                  dataTableRef?.current?.scrollIntoView({
                    behavior: 'smooth',
                    block: 'start',
                  });
                }, 200);
              }
            }}
          />

          <TimePeriodDataTable
            ref={dataTableRef}
            footnotesClassName="govuk-!-width-two-thirds"
            fullTable={table}
            query={query}
            tableHeadersConfig={tableHeaders}
            onError={message => {
              toggleHasTableError.on();
              logEvent({
                category: 'Table Tool',
                action: 'Table rendering error',
                label: message,
              });
            }}
          />
        </>
      )}
      {!hasTableError && (
        <>
          <div className="govuk-!-margin-bottom-7">
            <TableToolShare
              tableHeaders={tableHeaders}
              query={query}
              selectedPublication={selectedPublication}
              newPermalinks={newPermalinks}
            />
          </div>

          <DownloadTable
            fullTable={table}
            fileName={`data-${selectedPublication.slug}`}
            onCsvDownload={() => tableBuilderService.getTableCsv(query)}
            tableRef={dataTableRef}
            onSubmit={fileFormat =>
              logEvent({
                category: 'Table tool',
                action:
                  fileFormat === 'csv'
                    ? 'CSV download button clicked'
                    : 'ODS download button clicked',
                label: `${table.subjectMeta.publicationName} between ${
                  table.subjectMeta.timePeriodRange[0].label
                } and ${
                  table.subjectMeta.timePeriodRange[
                    table.subjectMeta.timePeriodRange.length - 1
                  ].label
                }`,
              })
            }
          />
        </>
      )}

      <TableToolInfo
        contactDetails={publication?.contact}
        methodologyLinks={getMethodologyLinks()}
        releaseLink={
          <>
            {selectedPublication.selectedRelease.latestData ? (
              <Link to={`/find-statistics/${selectedPublication.slug}`}>
                {`${selectedPublication.title}, ${selectedPublication.selectedRelease.title}`}
              </Link>
            ) : (
              <Link
                to={`/find-statistics/${selectedPublication.slug}/${selectedPublication.selectedRelease.slug}`}
              >
                {`${selectedPublication.title}, ${selectedPublication.selectedRelease.title}`}
              </Link>
            )}
          </>
        }
      />
    </div>
  );
};

export default memo(TableToolFinalStep);
