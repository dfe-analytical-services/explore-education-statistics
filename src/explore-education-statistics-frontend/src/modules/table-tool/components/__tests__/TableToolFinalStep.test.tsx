import TableToolFinalStep from '@frontend/modules/table-tool/components/TableToolFinalStep';
import { render, screen, waitFor } from '@testing-library/react';
import _publicationService from '@common/services/publicationService';
import {
  testTable,
  testTableHeaders,
  testQuery,
  testSelectedPublicationWithLatestRelease,
  testSelectedPublicationWithNonLatestRelease,
  testPublicationRelease,
} from '@frontend/modules/table-tool/components/__tests__/__data__/tableData';
import userEvent from '@testing-library/user-event';
import noop from 'lodash/noop';
import React from 'react';

jest.mock('@common/services/publicationService');

const publicationService = _publicationService as jest.Mocked<
  typeof _publicationService
>;

describe('TableToolFinalStep', () => {
  test('renders the final step successfully', async () => {
    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationRelease,
    );

    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    // test the reordering controls for the table headers are present
    expect(
      screen.getByRole('button', {
        name: 'Move and reorder table headers',
      }),
    ).toBeInTheDocument();

    // test that the preview table is rendered correctly
    expect(screen.getByTestId('dataTableCaption')).toHaveTextContent(
      "'dates' for 02/04/2021 and 02/06/2020 in England between 2020 Week 23 and 2020 Week 26",
    );

    const table = screen.queryByTestId('dataTableCaption-table');
    expect(table).toBeInTheDocument();
    expect(table).toMatchSnapshot();

    // test the permalink controls are present
    expect(
      screen.queryByRole('button', { name: 'Generate shareable link' }),
    ).toBeInTheDocument();

    // test that the related information is rendered correctly
    const relatedInfoRevealButton = screen.getByRole('button', {
      name: 'Related information',
    });

    expect(relatedInfoRevealButton).toBeInTheDocument();

    userEvent.click(relatedInfoRevealButton);

    await waitFor(() => {
      expect(
        screen.queryByRole('link', {
          name: 'Test publication, Latest Release Title',
        }),
      ).toBeInTheDocument();
      expect(
        screen.queryByRole('radio', {
          name: 'Table in ODS format (spreadsheet, with title and footnotes)',
        }),
      ).toBeInTheDocument();
      expect(
        screen.queryByRole('radio', {
          name: 'Table in CSV format (flat file, with location codes)',
        }),
      ).toBeInTheDocument();
      expect(
        screen.queryByRole('button', {
          name: 'Download table',
        }),
      ).toBeInTheDocument();
    });

    // test that contact us section is rendered correctly
    const contactUsRevealButton = screen.getByRole('button', {
      name: 'Contact us',
    });
    expect(contactUsRevealButton).toBeInTheDocument();

    userEvent.click(contactUsRevealButton);

    await waitFor(() => {
      expect(screen.queryByText('The team name')).toBeInTheDocument();
      expect(
        screen.queryByRole('link', {
          name: 'team@name.com',
        }),
      ).toBeInTheDocument();
    });
  });

  test('shows and hides table header reordering controls successfully', async () => {
    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationRelease,
    );

    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    const reorderRevealButton = screen.getByRole('button', {
      name: 'Move and reorder table headers',
    });

    expect(
      screen.queryByRole('button', { name: 'Update and view reordered table' }),
    ).not.toBeInTheDocument();

    userEvent.click(reorderRevealButton);

    await waitFor(() =>
      expect(
        screen.queryByRole('button', {
          name: 'Update and view reordered table',
        }),
      ).toBeInTheDocument(),
    );
  });

  test(`renders the 'View the release for this data' URL with the Release's slug, if the selected Release is not the latest for the Publication`, async () => {
    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationRelease,
    );
    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithNonLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    const relatedInfoRevealButton = screen.getByRole('button', {
      name: 'Related information',
    });

    userEvent.click(relatedInfoRevealButton);

    await waitFor(() => {
      const viewReleaseLink = screen.getByRole('link', {
        name: 'Test publication, Selected Release Title',
      }) as HTMLAnchorElement;
      expect(viewReleaseLink.href).toEqual(
        'http://localhost/find-statistics/test-publication/selected-release-slug',
      );
    });
  });

  test(`renders the 'View the release for this data' URL with only the Publication slug, if the selected Release is the latest Release for that Publication`, async () => {
    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationRelease,
    );
    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    const relatedInfoRevealButton = screen.getByRole('button', {
      name: 'Related information',
    });

    userEvent.click(relatedInfoRevealButton);

    await waitFor(() => {
      const viewReleaseLink = screen.getByRole('link', {
        name: 'Test publication, Latest Release Title',
      }) as HTMLAnchorElement;

      expect(viewReleaseLink.href).toEqual(
        'http://localhost/find-statistics/test-publication',
      );
    });
  });

  test('renders the Table Tool final step correctly when this is the latest data', async () => {
    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationRelease,
    );
    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    expect(screen.queryByText('This is the latest data')).toBeInTheDocument();
    expect(
      screen.queryByText('This data is not from the latest release'),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByTestId('View latest data link'),
    ).not.toBeInTheDocument();
  });

  test(`renders the Table Tool final step correctly if this is not the latest data`, async () => {
    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationRelease,
    );
    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithNonLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    expect(
      screen.queryByText('This is the latest data'),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText('This data is not from the latest release'),
    ).toBeInTheDocument();

    const latestDataLink = screen.queryByTestId(
      'View latest data link',
    ) as HTMLAnchorElement;
    expect(latestDataLink).toBeInTheDocument();
    expect(latestDataLink.href).toEqual(
      'http://localhost/find-statistics/test-publication',
    );
    expect(latestDataLink.text).toContain('View latest data');
    expect(latestDataLink.text).toContain('Latest Release Title');

    const relatedInfoRevealButton = screen.getByRole('button', {
      name: 'Related information',
    });

    userEvent.click(relatedInfoRevealButton);

    await waitFor(() => {
      const viewReleaseLink = screen.getByRole('link', {
        name: 'Test publication, Selected Release Title',
      }) as HTMLAnchorElement;

      expect(viewReleaseLink.href).toEqual(
        'http://localhost/find-statistics/test-publication/selected-release-slug',
      );
    });
  });

  test('renders the methodology link correctly', async () => {
    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationRelease,
    );
    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    const relatedInfoRevealButton = screen.getByRole('button', {
      name: 'Related information',
    });

    userEvent.click(relatedInfoRevealButton);

    await waitFor(() => {
      const methodologyLink = screen.getByRole('link', {
        name: 'methodology title',
      }) as HTMLAnchorElement;

      expect(methodologyLink.href).toEqual('http://localhost/methodology/m1');
    });
  });

  test('renders the external methodology link correctly', async () => {
    const testPublicationWithExternalMethodology = {
      ...testPublicationRelease,
      publication: {
        ...testPublicationRelease.publication,
        methodologies: [],
        externalMethodology: {
          url: 'http://somewhere.com',
          title: 'An external methodology',
        },
      },
    };

    publicationService.getLatestPublicationRelease.mockResolvedValue(
      testPublicationWithExternalMethodology,
    );
    render(
      <TableToolFinalStep
        query={testQuery}
        table={testTable}
        tableHeaders={testTableHeaders}
        selectedPublication={testSelectedPublicationWithLatestRelease}
        onReorderTableHeaders={noop}
      />,
    );

    const relatedInfoRevealButton = screen.getByRole('button', {
      name: 'Related information',
    });

    userEvent.click(relatedInfoRevealButton);

    await waitFor(() => {
      const methodologyLink = screen.getByRole('link', {
        name: 'An external methodology',
      }) as HTMLAnchorElement;

      expect(methodologyLink.href).toEqual('http://somewhere.com/');
    });
  });
});
