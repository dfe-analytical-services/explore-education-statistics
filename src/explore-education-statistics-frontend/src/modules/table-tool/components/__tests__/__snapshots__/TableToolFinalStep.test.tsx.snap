// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableToolFinalStep renders the final step successfully 1`] = `
<table
  aria-labelledby="dataTableCaption"
  class="govuk-table table"
  data-testid="dataTableCaption-table"
>
  <thead
    class="tableHead"
  >
    <tr>
      <td
        colspan="2"
        rowspan="1"
      />
      <th
        colspan="1"
        rowspan="1"
        scope="col"
      >
        02/06/2020
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th
        class=""
        colspan="1"
        rowspan="1"
        scope="rowgroup"
      >
        Number of open settings
      </th>
      <th
        class=""
        colspan="1"
        rowspan="1"
        scope="row"
      >
        2020 Week 23
      </th>
      <td
        class="govuk-table__cell--numeric"
      >
        22,500
      </td>
    </tr>
    <tr>
      <th
        class=""
        colspan="1"
        rowspan="1"
        scope="rowgroup"
      >
        Proportion of settings open
      </th>
      <th
        class=""
        colspan="1"
        rowspan="1"
        scope="row"
      >
        2020 Week 23
      </th>
      <td
        class="govuk-table__cell--numeric"
      >
        1%
      </td>
    </tr>
  </tbody>
</table>
`;
